src/runtime/HalideRuntime.h

```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
 extern int halide_copy_to_device(void *user_context, struct halide_buffer_t *buf,
                                  const struct halide_device_interface_t *device_interface);
 
+/** Copy data from one buffer to another. The buffers may have
+ * different shapes and sizes, but the destination buffer's shape must
+ * be contained within the source buffer's shape. The source data is
+ * pulled from either device or host memory on the source, depending
+ * on the dirty flags. host is preferred if both are valid. The
+ * dst_device_interface parameter controls the destination memory
+ * space. NULL means host memory. */
+extern int halide_buffer_copy(void *user_context, struct halide_buffer_t *src,
+                              const struct halide_device_interface_t *dst_device_interface,
+                              struct halide_buffer_t *dst);
+
+/** Give the destination buffer a device allocation which is an alias
+ * for the same coordinate range in the source buffer. Modifies the
+ * device, device_interface, and the device_dirty flag only. Only
+ * supported by some device APIs (others will return
+ * halide_error_code_device_crop_unsupported). Call
+ * halide_device_release_crop instead of halide_device_free to clean
+ * up resources associated with the cropped view. Do not call
+ * device_free on the source buffer while the destination buffer still
+ * lives.  Note that the two buffers do not share dirty flags, so care
+ * must be taken to update them together as needed. Note also that
```

Comment:
I'm a bit dubious about the last part. We will likely solve this in Halide::Runtime::Buffer by always translating the crop back into the parent's coordinates and cropping from there. (This is tricky in the general case as the mins could be reset on one of the buffers. One has to be able to determine the relative offset of the device memory references I think...) In any event, we should explicitly state that multiple crops can be taken from the same parent buffer.

If crop from a previous crop is not allowed on some architectures, I'm tempted to force all devices to make it an error for portability reasons. (Though realistically, I think we should investigate getting rid of the restriction.)